### PR TITLE
janus-pp-rec: support HEVC AP(aggregation packet)

### PR DESCRIPTION
**Pull request by nu774**
Added support for HEVC Aggregation Packets (AP), described in rfc7798.

### CODE REVIEW

File: postprocessing/pp-h265.c [hunk 1]
```diff
@@ -338,6 +338,62 @@ int janus_pp_h265_preprocess(FILE *file, janus_pp_frame_packet *list) {
...
+				uint16_t unit = 0;
+				memcpy(&unit, p, sizeof(uint16_t));
+				unit = ntohs(unit);
+				uint8_t fbit = (unit & 0x8000) >> 15;
+				uint8_t type = (unit & 0x7E00) >> 9;
+				uint8_t lid = (unit & 0x01F8) >> 3;
+				uint8_t tid = (unit & 0x0007);
+
+				switch(type) {
+				case 32:
```
Comment (lminiero) [inline postprocessing/pp-h265.c:359]: Code style: `case` and `default` parts should all be indented, not at the same level as `switch` (the same applies to both the `switch` statements you added).

File: postprocessing/pp-h265.c [hunk 2]
```diff
@@ -451,7 +507,37 @@ int janus_pp_h265_process(FILE *file, janus_pp_frame_packet *list, int *working)
 				memset(temp + 2, 0x01, 1);
 				frameLen += 3;
 			}
-			if(type == 49) {
+			else if(type == 48) {
```
Comment (lminiero) [inline postprocessing/pp-h265.c:510]: Code style: `else` should be on the same line as the previous closing bracket.

File: postprocessing/pp-h265.c [hunk 3]
```diff
@@ -451,7 +507,37 @@ int janus_pp_h265_process(FILE *file, janus_pp_frame_packet *list, int *working)
...
+				for (uint8_t *p = buffer + 2; p < end - 2; p += payload_len) {
+					payload_len = (p[0] << 8 | p[1]);
+					p += 2;
+					uint8_t *temp = received_frame + frameLen;
+					temp[0] = 0;
+					temp[1] = 0;
+					temp[2] = 1;
+					memcpy(temp + 3, p, payload_len);
+					switch(p[0] >> 1) {
+					case 32: case 33: case 34:
```
Comment (lminiero) [inline postprocessing/pp-h265.c:523]: Code style: apart from indentation (see above), the three `case` should each have their own line.

File: postprocessing/pp-h265.c [hunk 4]
```diff
@@ -451,7 +507,37 @@ int janus_pp_h265_process(FILE *file, janus_pp_frame_packet *list, int *working)
...
+						break;
+					}
+					frameLen += 3 + payload_len;
+				}
+				if(tmp->next == NULL || tmp->next->ts > tmp->ts)
+					break;
+				tmp = tmp->next;
+				continue;
+			}
+			else if(type == 49) {
```
Comment (lminiero) [inline postprocessing/pp-h265.c:540]: See above on `else` on the same line as the bracket.

**Review by lminiero — COMMENTED** [review]
Thanks! Is there any specific reason for the patch? Mostly I'm interested in understanding if there are other WebRTC clients that do H.265 out there besides Safari which this patch is needed for. If so, I guess whether the `janus_h265_is_keyframe` helper in `utils.c` needs to be updated as well, as otherwise we won't detect keyframes in the cases you're extending here.

I just added a few notes inline on code style.

**Comment by januscla** [comment]
Thanks for your contribution, @nu774! Please make sure you sign our [CLA](https://cla.conf.meetecho.com/janus), as it's a required step before we can merge this.

**Comment by nu774** [comment]
Fixed code style.

> Is there any specific reason for the patch? Mostly I'm interested in understanding if there are other WebRTC clients that do H.265 out there besides Safari which this patch is needed for. If so, I wonder whether the janus_h265_is_keyframe helper in utils.c needs to be updated as well, as otherwise we won't detect keyframes in the cases you're extending here.

Well actually, I'm trying to push streams from ffmpeg via simple RTSP-WebRTC gateway built with https://github.com/pion/webrtc and https://github.com/aler9/gortsplib .
In case of H264 it seems to work, but in case of HEVC it doesn't.
I'm trying to investigate why, and I noticed I cannot check recorded bitstream with janus-pp-rec due to lack of HEVC AP support. That's why I created this patch.

In other word, it doesn't have something to do with an existing implementation.
AP support is needed in my case because ffmpeg generates APs when re-packetizing HEVC short NAL units for RTP transport.

IMHO, RTP senders don't necessarily have to implement APs (it just becomes more redundant without them), but receivers should be able to handle them.

**Comment by lminiero** [comment]
Thanks for the clarification! I'll take note we may have to change the `janus_h265_is_keyframe` code in the future, but for now this looks good to merge for me :+1:

**Pull request merged**